### DISCUSSION TITLE
Common Critirea IPSec Requirement

### DISCUSSION THREAD

**Discussion by amdrsantos** [strongswan/strongswan#1073] [Q&A] [answered]
Part of the Common Criteria IPSec Requirement FCS_IPSEC_EXT.1.9, requires that the system shall support secret value "x" having a bit-length of at least twice the security strength of the negotiated Diffie-Hellman group.

strongSwan is using OpenSSL plugin:
`MODP_3072[openssl]
MODP_4096[openssl]
MODP_6144[openssl]
MODP_8192[openssl]
MODP_2048[openssl]
MODP_2048_224[openssl]
MODP_2048_256[openssl]
MODP_1536[openssl]
MODP_1024[openssl]
MODP_1024_160[openssl]
MODP_768[openssl]`

But how can I check if the secret value "x" (private key) has at least the bit-length of Diffie-Hellman ?

**Comment by tobiasbrunner** [comment, marked as answer]
By default, strongSwan uses a conservative approach and generates DH secrets with the same length as the DH group's prime. Only if `charon.dh_exponent_ansi_x9_42`  is disabled are the generated secrets shorter and optimized to the cryptographic strength of each DH group (e.g. for `mod2048` the secret will be 384 bits instead of 2048). These are based on the recommendations in [RFC 3526](https://datatracker.ietf.org/doc/html/rfc3526#section-8) (the values we use are still rather conservative), which are similar to those in NIST 800-57 so even the shorter lengths satisfy this CC requirement (e.g. for `modp2048` NIST 800-57 gives a security strength of 112 bits, so 384 bits is way above the required 224 bits).
